Re-implement the sidebar's vertical toolbar

robertknight opened the pull request:
This PR re-implements the vertical toolbar on the left edge of the sidebar
to make future changes to the toolbar UI easier (eg. some upcoming a11y changes)
and to decouple it from the rest of the annotator application (the toolbar and `Sidebar`/`Guest` classes currently access each other's internals in a rather haphazard way).

The new implementation has a similar code structure to the `Adder` toolbar that is
shown when selecting text. It consists of:

 1. A Preact component in `src/annotator/components/toolbar`
    which renders the toolbar UI according to the current state

 2. A controller class in `src/annotator/toolbar.js` which renders the
    Preact component into a container element and has provides properties
    that the `Sidebar` and `Guest` classes can use to update its state.

    The controller takes only the container element and callbacks
    associated with each button as inputs, so it no longer calls methods
    on the Guest/Sidebar.

    The sidebar and guest now update the toolbar UI by setting properties
    on the toolbar controller, rather than using a mixture of manual
    manipulation of the toolbar DOM and publishing events which get
    eventually translated to method calls on the toolbar.

If we convert the remainder of the sidebar UI from jQuery/manual DOM updates to Preact in future, then we can remove the toolbar controller entirely and just render the toolbar component from within the sidebar component.

There are **two small intentional UI changes** in this PR, otherwise it should be the same before:
 - The label for the sidebar toggle button was changed from "Toggle or resize sidebar" to "Show annotation sidebar" to make it clear a) _which sidebar_ it refers to (in case the page has its own) and b) to make it more obvious what the "selected" state (ie. `aria-pressed="true"`) means
 - The label for the toggle highlights button was changed to "Show highlights" for similar reasons

lyzadanger commented on src/annotator/components/test/toolbar-test.js:
```diff
@@ -0,0 +1,82 @@
+  const createToolbar = props =>
+    mount(
+      <Toolbar
+        closeSidebar={noop}
```
I like your `noop` approach.

lyzadanger commented on src/annotator/components/test/toolbar-test.js:
```diff
@@ -0,0 +1,82 @@
+    assert.isFalse(wrapper.find('button').exists());
+  });
+
+  it('renders only the Close button if `useMinimalControls` is true', () => {
```
Does this need a clause about "and the sidebar is open"?

lyzadanger commented on src/annotator/components/test/toolbar-test.js:
```diff
@@ -0,0 +1,82 @@
+    assert.isTrue(findButton(wrapper, 'New page note').exists());
+  });
+
+  it('shows the "New annotation" button if `newAnnotationType` is `annotation`', () => {
```
HOLY MACKEREL: I never noticed that if you select text, the page note button turns into an annotation button before!!

lyzadanger commented on src/annotator/components/toolbar.js:
```diff
@@ -0,0 +1,135 @@
+  onClick,
+  selected,
+}) {
+  const handleClick = event => {
```
Can you elaborate on why we need to snarf up events here?

robertknight replied:
Yes. The sidebar has document-level event handlers that close the sidebar when clicks happen in the document. In some cases third-party code in the page may also handle clicks in some way. See also https://github.com/hypothesis/client/pull/2106.

lyzadanger commented on src/annotator/components/toolbar.js:
```diff
@@ -0,0 +1,135 @@
+  return (
+    <button
+      // This currently uses the icon font. This needs to be converted to
+      // `SvgIcon`.
```
Are we prevented from doing this (now) by the absence of these icons in SVG format, or by some other technical impediment?

robertknight replied:
We don't have all these icons in SVG format already in the repo, so deferred this to follow-up work.

lyzadanger commented on src/annotator/components/toolbar.js:
```diff
@@ -0,0 +1,135 @@
+  useMinimalControls = false,
+}) {
+  return (
+    <div>
```
This `div` doesn't need a `role` or anything, does it? We've kind of got a set of controls here, so thought I'd ask...

robertknight replied:
We could make this an ARIA toolbar. If we do that though, we need to implement the correct keyboard interaction behavior at the same time.

lyzadanger commented on src/annotator/sidebar.coffee:
```diff
@@ -16,38 +19,51 @@ module.exports = class Sidebar extends Host
+          this.hide()
+      setHighlightsVisible: (show) => this.setAllVisibleHighlights(show)
+    })
+    @toolbar.useMinimalControls = config.theme == 'clean'
```
Just wondering if config could be accessed from within the toolbar itself instead of being passed in?

robertknight replied:
I'm not quite sure I've understood correctly here, but the way I've set this PR up, the toolbar widget doesn't receive the global configuration options any more. Instead the sidebar is responsible for mapping global config options to toolbar config.

lyzadanger commented on src/annotator/test/sidebar-test.coffee:
```diff
@@ -331,6 +365,12 @@ describe 'Sidebar', ->
       sidebar.show()
       assert.isFalse sidebar.visibleHighlights
 
+    it 'updates the toolbar', ->
```
Could this test description (and an identical one for `#hide`) be a little clearer?

lyzadanger commented on src/annotator/test/toolbar-test.js:
```diff
@@ -0,0 +1,149 @@
+    assert.equal(controller.newAnnotationType, 'note');
+  });
+
+  it('re-renders when `useMinimalControls` changes', () => {
```
It's so nice that we get all of this for free with `preact` components :)

lyzadanger commented on src/annotator/test/toolbar-test.js:
```diff
@@ -0,0 +1,149 @@
+    });
+  });
+
+  it('toggles sidebar visibility', () => {
```
The `it`s in this and the following couple of tests are ambiguous to me.

lyzadanger commented on src/annotator/components/toolbar.js:
```diff
@@ -0,0 +1,135 @@
+  closeSidebar: propTypes.func.isRequired,
+
+  /** Callback for the "Create page note" button. */
+  createAnnotation: propTypes.func.isRequired,
```
This comment is invalid now that I understand what this actually _does_.

robertknight commented on src/annotator/components/toolbar.js:
```diff
@@ -0,0 +1,135 @@
+}
+
+ToolbarButton.propTypes = {
+  buttonRef: propTypes.any,
```
In this context I don't think I can add value over the JSX a few lines up. If there is anything that isn't clear, please do say though.

robertknight commented on src/annotator/sidebar.coffee:
```diff
@@ -16,38 +19,51 @@ module.exports = class Sidebar extends Host
     if config.theme == 'clean' || config.externalContainerSelector
       delete config.pluginClasses.BucketBar
+
+    # TODO - Make this work again.
```
This comment is obsolete. I had forgotten to remove it. There is an `if @frame` block a few lines below which handles this now.

In short, the publisher can provide their own container element for the sidebar, in which case we won't show the vertical toolbar or bucket bar.

robertknight commented:
Thanks for the feedback @lyzadanger. I have updated this PR with improved test descriptions and some additional comments. I also removed the obsolete "TODO".
